doc/design/publish-1.md

```diff
@@ -20,29 +25,26 @@ Each Asset could have a different metadata record for the same AssetBlob.
 
 Whenever a new AssetBlob is created, a background task runs to compute the sha256 checksum (a "digest") of the blob.
 This can take a long time for large blobs, as it must stream all of the data from S3.
+This checksum is stored on the AssetBlob, not in the Asset metadata.
 
 ### Asset
-Assets consist of a JSON blob of metadata and a reference to an AssetBlob.
+Assets consist of a JSON blob of Asset metadata and a reference to an AssetBlob.
 Versions and Assets are many-to-many; a Version can have any number of Assets, and an Asset can belong to any number of Versions (even, hypothetically, versions in different Dandisets).
 Assets also track their validation status.
 
 Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
 Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
 This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.
 
-Similarly to Versions, there are two kinds of Asset: draft and published.
-Draft Assets are only allowed to belong to draft Versions.
-Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).
-
-Note that published Assets can belong to both draft and published Versions.
-
-When a published Asset is modified in this way, the publish metadata is stripped out, effectively creating a new draft Asset.
+When Asset metadata is requested through the API, some extra information is injected.
+* `size`, `path` are both properties of the AssetBlob, so these are injected.
+* `contentUrl` is dependent on which version the Version the asset belongs to, so it must be injected.
+* If the AssetBlob digest has finished calculating, it is injected into the `digests` field of the metadata.
+* The oldest published Version that the Asset belongs to is queried, and the `publishBy` and `datePublished` fields for that Version are injected into the Asset metadata.
+If an asset does not belong to any published Versions, `publishedBy` is not injected.
 
-_In general_, Assets should be immutable; any modification will instead create a new draft Asset.
-* One exception to this is calculating the AssetBlob digest.
-When the digest completes, it will be injected directly into the metadata of any Assets referencing that AssetBlob.
-* The other exception is creating a published Asset during the publish process.
-This involves directly injecting the new metadata fields into the draft Asset, effectively turning it into a published Asset in place.
+Because "modifying" an Asset actually creates a new Asset, Asset metadata will largely be stable despite having dynamically computed fields.
+`digests` will change when the digest is calculated, but everything else is immutable.
```

Comment:
Why would a digest of an otherwise-immutable data collection change?

Reply:
Immediately after a large new blob is upload and an asset is created with it, the digest is unknown because it hasn't been calculated. However, the metadata still exists and should be exposed in the API, so we have to provide it with no digest. When the digest calculation completes, it should be reported, so it is injected into the `digests` field of the metadata.

Reply:
just a point of clarification, it should have one digest immediately on asset creation the `dandi-etag`.

Reply:
So (roughly speaking) `digests` is "mutable" just in the narrow sense that it will receive a new value (in addition to the dandi-etag) when the digest is computed, but it's not that it will change value at other times.

Thanks for clarifiying.